### PR TITLE
bep: add user notification settings plan

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

Notification BEP addition for user specific notification settings

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: beps/0001-notifications-system/README.md [hunk 1]
```diff
@@ -380,10 +380,39 @@ interface SignalApi {
...
 
+### User specific notification settings
+
+To allow the users more fine-grained control of notifications, user must have an option to unsubscribe from specific origins, topics and channels.
+
+By default, all notifications are enabled for the user in all notification channels. Frontend plugin provides an user interface to change these settings for example as described in the following image:
+
+![user notification settings UI example](./UserNotificationSettings.png)
+
+Each notification processor with `send` functionality will be listed as a separate channel among the Backstage internal notification channel `Web`. For this, each notification processor must implement a function to get a human readable name to be rendered in the frontend.
```
Comment (mareklibra) [inline beps/0001-notifications-system/README.md:394]: I assume that the `Web` channel filter will be a processor as well. If so, I would emphasize it in the text.

That web-processor can be the first&default one in the array of the `NotificationsProcessingExtensionPoint`. Do we need to provide a way how to disable it?

Probably the easiest way is to implement that "web-processor" in the notifications-backend.
But doing it in a brand new plugin will provide a nice example for the authors of future processors. Wdyt?

File: beps/0001-notifications-system/README.md [hunk 2]
```diff
@@ -380,10 +380,39 @@ interface SignalApi {
...
+
+```ts
+export type NotificationSetting = {
+  origin: string;
+  topic?: string;
+  channels: Record<string, boolean>;
+};
+
+export type UserNotificationSettings = {
+  settings: NotificationSetting[];
```
Comment (mareklibra) [inline beps/0001-notifications-system/README.md:410]: What about
```suggestion
  routes: NotificationSetting[];
```
Or another variant on `userNotificaionRoutes`?

There can be additional settings needed in the future, i.e. preselection of filters.

**Review by mareklibra — COMMENTED** [review]
Any proposal about UXD?

File: beps/0001-notifications-system/README.md (hunk 1 above)
Comment (drodil) [reply beps/0001-notifications-system/README.md:394]: Web is not a processor. It's the current way notifications work in the Backstage. Basically this means users can disable notifications in the backstage UI but still receive them through specific processors such as email.

File: beps/0001-notifications-system/README.md (hunk 2 above)
Comment (drodil) [reply beps/0001-notifications-system/README.md:410]: Don't understand, could you elaborate?

File: beps/0001-notifications-system/README.md (hunk 1 above)
Comment (mareklibra) [reply beps/0001-notifications-system/README.md:394]: Agree. But the described filtering that notifications of particular channel should not be delivered to the Web can be handled via a new processor which can call itself as the `Web`

Comment (drodil) [reply beps/0001-notifications-system/README.md:394]: Thinking about it, it could be possible to separate it as own processor. It should be in own module I think so that users can disable it just by removing the dependency to that module. This though would be a breaking change as there would not be any notification support from the backend alone. Any opinions @Rugvib?

Comment (drodil) [reply beps/0001-notifications-system/README.md:394]: Also this would mean quite a big refactoring to move also the database to the web processor module. But it might still be a good idea just to keep the notifications as modular as possible

File: beps/0001-notifications-system/README.md (hunk 2 above)
Comment (mareklibra) [reply beps/0001-notifications-system/README.md:410]: I am pointing to a way how we persist/transfer all settings related to the Notifications in an extensible way.

The so far proposed settings are in fact routing - whether a processor will handle the particular message.

In the (near-) future, we can go beyond that settings. I.e. we can add a way how the user can persist his selection of filters on the NotificationsPage and so create "a personalized view". Or something else. The `UserNotificationSettings` is a nice wrapper for all these user-specific settings. 

In other words, the proposed `UserNotificationSettings` has the only property called `settings`. Any future additions will be kind of `settings` as well, just for something else.

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
Not stale, we need some solution for this as new processors keep popping

**Review by Rugvip — APPROVED** [review]
👍, I think it's a solid idea as such but we can probably iterate on the naming of things a bit. Gonna merge so that we have this to look at in the BEP rather than hanging in a PR
